web/frontend/src/components/slam.vue

```diff
@@ -1,193 +1,208 @@
+    length += array.length;
+  }
+  return result;
 };
 
-/*
- * const fetchSLAMMap = (name: string) => {
- *   return nextTick(() => {
- *     const req = new slamApi.GetPointCloudMapStreamRequest();
- *     req.setName(name);
- *     rcLogConditionally(req);
- *     const chunks: Uint8Array[] = [];
- *     const getPointCloudMapStream = props.client.slamService.getPointCloudMapStream(req);
- *     getPointCloudMapStream.on('data', (res) => {
- *       const chunk = res.getPointCloudPcdChunk_asU8();
- *       chunks.push(chunk);
- *     });
- *     getPointCloudMapStream.on('status', (status) => {
- *       if (status.code !== 0) {
- *         const error = {
- *           message: status.details,
- *           code: status.code,
- *           metadata: status.metadata,
- *         };
- *         displayError(error);
- *       }
- *     });
- *     getPointCloudMapStream.on('end', (end) => {
- *       if (end === undefined || end.code !== 0) {
- *         // the error will be logged in the 'status' callback
- *         return;
- *       }
- *       pointcloud = concatArrayU8(chunks);
- *       pointCloudUpdateCount += 1;
- *     });
- *   });
- * };
- */
+const fetchSLAMMap = (name: string): Promise<Uint8Array> => {
+  return new Promise((resolve, reject) => {
+    const req = new slamApi.GetPointCloudMapStreamRequest();
+    req.setName(name);
+    rcLogConditionally(req);
+    const chunks: Uint8Array[] = [];
 
-/*
- * const fetchSLAMPose = (name: string) => {
- *   return nextTick(() => {
- *     const req = new slamApi.GetPositionNewRequest();
- *     req.setName(name);
- *     props.client.slamService.getPositionNew(req, new grpc.Metadata(), (error, res): void => {
- *       if (error) {
- *         displayError(error);
- *         return;
- *       }
- *       pose = res!.getPose()!;
- *     });
- *   });
- * };
- */
+    const getPointCloudMapStream = props.client.slamService.getPointCloudMapStream(req);
+    getPointCloudMapStream.on('data', (res) => {
+      const chunk = res.getPointCloudPcdChunk_asU8();
+      chunks.push(chunk);
+    });
+    getPointCloudMapStream.on('status', (status) => {
+      if (status.code !== 0) {
+        const error = {
+          message: status.details,
+          code: status.code,
+          metadata: status.metadata,
+        };
+        reject(error);
+      }
+    });
+    getPointCloudMapStream.on('end', (end) => {
+      if (end === undefined || end.code !== 0) {
+        // the error will be logged in the 'status' callback
+        return;
+      }
+      const arr = concatArrayU8(chunks);
+      resolve(arr);
+    });
+  });
+};
+
+const fetchSLAMPose = (name: string): Promise<commonApi.Pose> => {
+  return new Promise((resolve, reject): void => {
+    const req = new slamApi.GetPositionNewRequest();
+    req.setName(name);
+    props.client.slamService.getPositionNew(req, new grpc.Metadata(), (error, res): void => {
+      if (error) {
+        reject(error);
+        return;
```

Comment:
[q] why does this have a return; when the reject in the fetchSLAMMap doesn't? (note: may be a lack of understand on my part regarding certain js functionalities)

Reply:
reject is a function call, once it returns execution will pick up on line 86 & as a result if we don't want to then call resolve after calling reject, we need to early return.

Let me know if I got any of that wrong @micheal-parks

Reply:
Correct

Reply:
Got it, and because when getting data from the stream in fetchSLAMMap, we don't mind if an error causes us to lose a chunk, we can get the rest of the data.